src/pybind/mgr/dashboard/HACKING.rst

```diff
@@ -2052,6 +2053,22 @@ example below:
     '400':{'my_response': (str, 'Description of my_response')}})
   def method(): pass
 
+RESTful API YML Generation via `tox`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While the primary purpose of the introduction of ``tox``-based yml generation
+is for use in the ``build-doc`` script located in the ``\admin\`` directory,
+it is possible to use ``tox`` directly to run the ``docs.py`` script. This is
+done by way of the following syntax:
+
+.. code-block::  
+
+  tox -e apidocs <filename.yml>
+
+It will be generated in the ``src/pybind/mgr/`` folder as the specified file
+name. It is worth noting that only absolute paths are supported by this
+function at this time. 
```

Comment:
Why should the dashboard API be created in `src/pybind/mgr` and not in `src/pybind/mgr/dashboard`? As said earlier it's not normal to search for the filename I provided to a command to search one level up.

It's also worth a lot that at least absolute paths work as expected. As they work please update the description to `absolute_path_to_output_file` or something similar (for example `$PWD/filename.yml`).

Just providing the filename results in a bug as the behavior is not expected.

Reply:
> Why should the dashboard API be created in src/pybind/mgr and not in src/pybind/mgr/dashboard? As said earlier it's not normal to search for the filename I provided to a command to search one level up.

@Devp00l Yes, it is normal because `dashboard` is a package, not a regular script. All our python files use relative imports and behave as submodules of this `dashboard` package, so the only way to invoke an independent file is doing so as a submodule of that package, like `python3 -m dashboard.controllers.docs {posargs}`. If you try to do `python3 -m controllers.docs`  from `mgr/dashboard`, you'll get `ValueError: attempted relative import beyond top-level package`. And you if you do `python3 -m ..dashboard.controllers.docs` you'll get `Relative module names not supported`.

Reply:
@CourtneyCCaldwell @Devp00l one thing we can do try `PYTHONPATH=$PYTHONPATH:.. python3 -m dashboard.controllers.docs` from `..mgr/dashboard` directory. It's hacky but it might work.

Reply:
> > Why should the dashboard API be created in src/pybind/mgr and not in src/pybind/mgr/dashboard? As said earlier it's not normal to search for the filename I provided to a command to search one level up.
> 
> @Devp00l Yes, it is normal because `dashboard` is a package, not a regular script. All our python files use relative imports and behave as submodules of this `dashboard` package, so the only way to invoke an independent file is doing so as a submodule of that package, like `python3 -m dashboard.controllers.docs {posargs}`. If you try to do `python3 -m controllers.docs` from `mgr/dashboard`, you'll get `ValueError: attempted relative import beyond top-level package`. And you if you do `python3 -m ..dashboard.controllers.docs` you'll get `Relative module names not supported`.

I mend from the end user experience ;) I know that the file is placed one level up through the `changedir` usage in `tox.ini`, but this is not the expected end user experience. If I would call some program on command line with an filename as argument where something should be saved, and the doc said I can and don't need an absolute path, and the new file don't ends up in the current directory, I would purge the program and search for an alternative.
As said before it would be fine with me if we say only absolute paths are supported, but currently the opposite is done.

Reply:
@Devp00l I found a fix for this (TL;TR by removing first entry in `sys.path`), but it's really hacky, and the ultimate cause for this is the messy directory structure in ceph-mgr modules: there's a `mgr/dashboard/tests` dir, and also a `mgr/tests` one, and we somehow expect that `from tests ...`  resolves sometimes to either one.

So my suggestion to @CourtneyCCaldwell is to keep the behaviour (`dashboard.controllers.docs` launched from `dashboard/..`) and force YAML path to be absolute.